### ISSUE TITLE
In queries, account names cannot contain non-ASCII characters followed by a whitespace

### ISSUE THREAD

**Issue by schoettl** [simonmichael/hledger#1374]
**Platform**: GNU/Linux
**Hledger version**: 1.19.99 (compiled from source yesterday)
**Ledger file encoding**: UTF-8

## Problem

hledger command line **queries with an account** specification (in quotes) do **not handle non-ASCII characters** when the word is followed by a whitespace.

## Example

```
$ hledger accounts | grep Öffentl
Income:Öffentliche Zuschüsse
```

```
$ hledger register "Income:Öffentliche"
...
(register entries)
...
$ hledger register "Income:Öffentliche "
$ hledger register "Income:Öffentliche Zusch"
$
```
The first one works, the second and third don't (no output). Same for `hledger balance` command.

## Notes

- Queries with ASCII-only account names and spaces work.
- The bug wasn't there in January 2020.

**Comment by schoettl** [comment]
I'm git bisecting right now.

**Comment by schoettl** [comment]
Got it, it's commit 7d1e6d7d12f3c3a042661c1d6963523b5d5f676b by @Xitian9 

``` haskell
-- | Double-quote this string if it contains whitespace, single quotes
-- or double-quotes, escaping the quotes as needed.
quoteIfNeeded :: String -> String
quoteIfNeeded s | any (`elem` s) (quotechars++whitespacechars++redirectchars) = show s
                | otherwise = s
```
The `show s` is new and a cool idea but it also backslash-escapes non-ASCII characters. That's why they don't match anymore. Can maybe `printf` do some proper quoting? Like in bash: `printf "%q\n" "Ökonom followed by space"`

**Comment by schoettl** [comment]
I would reintroduce and use this function:
``` haskell
escapeDoubleQuotes :: String -> String
escapeDoubleQuotes = ...
```

[`subRegex`](https://hackage.haskell.org/package/regex-compat-0.95.2.0/docs/Text-Regex.html) from regex-compat could do the job. Or [text-regex-replace](https://hackage.haskell.org/package/text-regex-replace-0.1.1.3/docs/Data-Text-ICU-Replace.html). Both packages are not yet dependencies.

**Comment by Xitian9** [comment]
This is a duplicate of #1364, for which there is a pending fix in #1367. Behaviour with respect to string quoting was never correct, and has at various points been wrong in different ways, see e.g. #1368.

My opinion is that we're trying to do too much work with quotes, getting around limitations of using the wrong data structure. The flow of things is this:

1. Command line gives a list of strings
2. Combine those strings by quoting each string if necessary and then concatenating them. This requires escaping the inner quotation marks so they don't interefere with the outer quotation marks.
3. Parse the resulting string into a query.

Probably the better option is to omit step 2, and deal directly with the list of strings.

**Comment by Xitian9** [comment]
If you look carefully at the old definition of `escapeDoubleQuotes`, you'll see that it doesn't actually do anything at all: no quotes are escaped, and all that happens in `quoteIfNeeded` is quotes are added. This leads to bad behaviour when there are quotes in the string itself.

**Comment by simonmichael** [comment]
Fixed by #1367, hopefully.

**Comment by schoettl** [comment]
Thanks!
